discovery/dockerswarm/services.go

```diff
@@ -0,0 +1,136 @@
+	tg := &targetgroup.Group{
+		Source: "DockerSwarm",
+	}
+
+	services, err := d.client.ServiceList(ctx, types.ServiceListOptions{Filters: makeArgs(d.filters)})
+	if err != nil {
+		return nil, fmt.Errorf("error while listing swarm services: %w", err)
+	}
+
+	networkLabels := make(map[string]model.LabelSet, 1)
+
+	for _, s := range services {
+		for _, e := range s.Endpoint.Ports {
+			if e.Protocol != swarm.PortConfigProtocolTCP {
+				continue
+			}
+			for _, p := range s.Endpoint.VirtualIPs {
+				labels := model.LabelSet{
+					swarmLabelServiceEndpointPortName:        model.LabelValue(e.Name),
+					swarmLabelServiceEndpointPortPublishMode: model.LabelValue(e.PublishMode),
+					swarmLabelServiceID:                      model.LabelValue(s.ID),
+					swarmLabelServiceMode:                    getServiceValueMode(s),
+					swarmLabelServiceName:                    model.LabelValue(s.Spec.Name),
+					swarmLabelServiceTaskContainerHostname:   model.LabelValue(s.Spec.TaskTemplate.ContainerSpec.Hostname),
+					swarmLabelServiceTaskContainerImage:      model.LabelValue(s.Spec.TaskTemplate.ContainerSpec.Image),
+				}
+
+				if s.UpdateStatus != nil {
+					labels[swarmLabelServiceUpdatingStatus] = model.LabelValue(s.UpdateStatus.State)
+				}
+
+				for k, v := range s.Spec.Labels {
+					ln := strutil.SanitizeLabelName(k)
+					labels[model.LabelName(swarmLabelServiceLabelPrefix+ln)] = model.LabelValue(v)
+				}
+
+				ip, _, err := net.ParseCIDR(p.Addr)
+				if err != nil {
+					return nil, fmt.Errorf("error while parsing address %s: %w", p.Addr, err)
+				}
+				addr := net.JoinHostPort(ip.String(), strconv.FormatUint(uint64(e.PublishedPort), 10))
+				labels[model.AddressLabel] = model.LabelValue(addr)
+
+				if _, ok := networkLabels[p.NetworkID]; !ok {
+					lbs, err := d.getNetworkLabels(ctx, p.NetworkID)
+					if err != nil {
+						return nil, fmt.Errorf("error while inspecting network %s: %w", p.NetworkID, err)
+					}
+					networkLabels[p.NetworkID] = lbs
+				}
+
+				for k, v := range networkLabels[p.NetworkID] {
+					labels[k] = v
+				}
+
+				tg.Targets = append(tg.Targets, labels)
+			}
+		}
+	}
+	return []*targetgroup.Group{tg}, nil
+}
+
+func (d *Discovery) getServiceLabelsAndPorts(ctx context.Context, serviceID string) (model.LabelSet, []swarm.PortConfig, error) {
+	s, _, err := d.client.ServiceInspectWithRaw(ctx, serviceID, types.ServiceInspectOptions{})
+	if err != nil {
+		return nil, nil, err
+	}
+	labels := model.LabelSet{
+		swarmLabelServiceID:   model.LabelValue(s.ID),
+		swarmLabelServiceName: model.LabelValue(s.Spec.Name),
+		swarmLabelServiceMode: getServiceValueMode(s),
+	}
+	for k, v := range s.Spec.Labels {
+		ln := strutil.SanitizeLabelName(k)
+		labels[model.LabelName(swarmLabelServiceLabelPrefix+ln)] = model.LabelValue(v)
+	}
+
+	return labels, s.Endpoint.Ports, nil
+}
+
+func getServiceValueMode(s swarm.Service) model.LabelValue {
+	if s.Spec.Mode.Global != nil {
+		return swarmLabelValueModeGlobal
```

Comment:
I don't like indirection like this so I have to search the whole codebase to see what the value actually is, use a constant here.